pkg/traefik/traefik.go

```diff
@@ -0,0 +1,310 @@
+}
+
+func (t *Traefik) URL(ctx context.Context, prefix string) (string, error) {
+	if t.endpoint == "" {
+		var err error
+		if t.endpoint, err = t.Endpoint(ctx); err != nil {
+			return "", ErrTraefikIPNotFound.Wrap(err)
+		}
+	}
+	return fmt.Sprintf("http://%s/%s", t.endpoint, prefix), nil
+}
+
+func (t *Traefik) Endpoint(ctx context.Context) (string, error) {
+	if t.K8s == nil {
+		return "", ErrTraefikClientNotInitialized
+	}
+	return t.K8s.GetServiceEndpoint(ctx, traefikServiceName)
+}
+
+func (t *Traefik) AddHost(ctx context.Context, serviceName, prefix string, portTCP int) error {
+	middlewareName, err := names.NewRandomK8("strip-" + prefix)
+	if err != nil {
+		return ErrGeneratingRandomK8sName.Wrap(err)
+	}
+
+	// middleware is required to strip the prefix from the service name
+	if err := t.createMiddleware(ctx, prefix, middlewareName); err != nil {
+		return err
+	}
+
+	return t.createIngressRoute(ctx, serviceName, prefix, middlewareName, portTCP)
+}
+
+// TODO: need to update the k8s pkg to handle service creation in more custom way
+func (t *Traefik) createService(ctx context.Context) error {
+	sCli := t.K8s.Clientset().CoreV1().Services(t.K8s.Namespace())
+
+	srv := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      traefikServiceName,
+			Namespace: t.K8s.Namespace(),
+			Labels:    map[string]string{appLabel: appLabelValue},
+		},
+		Spec: v1.ServiceSpec{
+			Selector: map[string]string{appLabel: appLabelValue},
+			Ports: []v1.ServicePort{
+				{
+					Name:       "web",
+					Protocol:   v1.ProtocolTCP,
+					Port:       Port,
+					TargetPort: intstr.FromInt(Port),
+				},
+				{
+					Name:       "websecure",
+					Protocol:   v1.ProtocolTCP,
+					Port:       PortSecure,
+					TargetPort: intstr.FromInt(PortSecure),
+				},
+			},
+			Type: v1.ServiceTypeLoadBalancer,
+		},
+	}
+
+	if _, err := sCli.Create(ctx, srv, metav1.CreateOptions{}); err != nil {
+		return ErrTraefikFailedToCreateService.Wrap(err)
+	}
+
+	logrus.Debugf("Service %s created successfully.", traefikServiceName)
+	return nil
+}
+
+func (t *Traefik) createMiddleware(ctx context.Context, serviceName, middlewareName string) error {
+	middleware := &unstructured.Unstructured{
+		Object: map[string]interface{}{
+			"apiVersion": "traefik.io/v1alpha1",
+			"kind":       "Middleware",
+			"metadata": map[string]interface{}{
+				"name":      middlewareName,
+				"namespace": t.K8s.Namespace(),
+			},
+			"spec": map[string]interface{}{
+				"stripPrefix": map[string]interface{}{
+					"prefixes": []string{"/" + serviceName},
+				},
+			},
+		},
+	}
+
+	middlewareResource := schema.GroupVersionResource{
+		Group:    "traefik.io",
+		Version:  "v1alpha1",
+		Resource: "middlewares",
+	}
+
+	_, err := t.K8s.DynamicClient().Resource(middlewareResource).Namespace(t.K8s.Namespace()).Create(ctx, middleware, metav1.CreateOptions{})
+	if err != nil {
+		return ErrTraefikMiddlewareCreationFailed.Wrap(err)
+	}
+	return nil
+}
```

Comment:
Open thought/question:

Knuu was designed with the idea that the Kubernetes Cluster does not need any specific setup (Like installing operators). So you could spin up a new Cluster and run your tests.
For example, chaos-mesh was requested but required an operator, so I created a separate repository. If the dev wants to use chaos-mesh, he/she can import that 'add-on' to the knuu Instance and control the traffic.

I'm not sure how to handle that reverse proxy topic.

Please share your thoughts on that topic @mojtaba-esk @jrmanes

Reply:
Good question. We can go either with the add on approach or since right now it is used by the knuu instance when BitTwister is activated or we might add some more fucntions in future like running a custom code, etc... another option is to keep it as a configurable variable so user can activate it when they need it.